cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaNodeSetController.java

```diff
@@ -0,0 +1,478 @@
+        }
+    }
+
+    /**
+     * Checks if the Pod which should be enqueued belongs to a KafkaNodeSet this controller manages and whether the
+     * Kafka cluster which owns it matches the CR selector. If it does, it will enqueue the reconciliation. This is used
+     * to enqueue reconciliations based on Pod events.
+     *
+     * Note: The reconciliation is enqueued per KafkaNodeSet to which the pod belongs and not based on the Pod itself.
+     *
+     * @param pod      Pod which should be checked and possibly enqueued
+     * @param action   The action from the event which triggered this
+     */
+    private void enqueuePod(Pod pod, String action) {
+        LOGGER.debugOp("Pod {} in namespace {} was {}", pod.getMetadata().getName(), pod.getMetadata().getNamespace(), action);
+
+        KafkaNodeSet parentNodeSet = kafkaNodeSetLister
+                .namespace(pod.getMetadata().getNamespace())
+                .list()
+                .stream()
+                .filter(nodeSet -> nodeSet.getSpec() != null
+                        && Util.matchesSelector(Optional.ofNullable(nodeSet.getSpec().getSelector()), pod))
+                .findFirst().orElse(null);
+
+        if (parentNodeSet != null) {
+            if (matchesCrSelector(parentNodeSet)) {
+                enqueue(new SimplifiedReconciliation(parentNodeSet.getMetadata().getNamespace(), parentNodeSet.getMetadata().getName()));
+            } else {
+                LOGGER.debugOp("Pod {} in namespace {} was {} but does not belong to a Kafka cluster managed by this operator", pod.getMetadata().getName(), pod.getMetadata().getNamespace(), action);
+            }
+        } else {
+            LOGGER.debugOp("Pod {} in namespace {} which was {} does not seem to be controlled by any KafkaNodeSet and will be ignored", pod.getMetadata().getName(), pod.getMetadata().getNamespace(), action);
+        }
+    }
+
+    /**
+     * Utility method which tries to find the Kafka cluster to which this KafkaNodeSet belongs and checks whether the CR
+     * selector labels match or not. The controller handles only Kafka clusters matching the CR selector labels. Other
+     * clusters are ignored
+     *
+     * @param nodeSet   KafkaNodeSet which should be checked
+     *
+     * @return          True if the KafkaNodeSet's Kafka cluster matches the selector labels
+     */
+    private boolean matchesCrSelector(KafkaNodeSet nodeSet)    {
+        if (nodeSet.getMetadata().getLabels() != null
+                && nodeSet.getMetadata().getLabels().get(Labels.STRIMZI_CLUSTER_LABEL) != null) {
+            // We find the matching Kafka cluster and check the CR selector
+            String kafkaClusterName = nodeSet.getMetadata().getLabels().get(Labels.STRIMZI_CLUSTER_LABEL);
+            Kafka kafka = kafkaLister.namespace(nodeSet.getMetadata().getNamespace()).get(kafkaClusterName);
```

Comment:
AFAICS this doesn't really depend on this resource being a `Kafka` CR. Any resource could be used, it's just we only care about the Kafka case, right?

Reply:
Well, right now, the parent of the KafkaNodeSet is always Kafka CR and that is where we need to check the label selector. We do not use it with any other resources. If we one day use it also with Connect or Bridge etc., we would need to extend this. But not sure we need to invest into that now.